pkg/yurttunnel/pki/certmanager/certmanager.go

```diff
@@ -17,21 +17,345 @@ limitations under the License.
+	certificateStore, err :=
+		certificate.NewFileStore(componentName, certDir, certDir, "", "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the server certificate store: %v", err)
+	}
+
+	getTemplate := func() *x509.CertificateRequest {
+		return &x509.CertificateRequest{
+			Subject: pkix.Name{
+				CommonName:   constants.YurttunneServerCSRCN,
+				Organization: organizations,
+			},
+			DNSNames:    dnsNames,
+			IPAddresses: ipAddrs,
+		}
+	}
+
+	serverCertManager, err := certificate.NewManager(&certificate.Config{
+		ClientFn: func(current *tls.Certificate) (clicert.CertificateSigningRequestInterface, error) {
+			return clientset.CertificatesV1beta1().CertificateSigningRequests(), nil
+		},
+		GetTemplate: getTemplate,
+		Usages: []certificates.KeyUsage{
+			certificates.UsageAny,
+		},
+		CertificateStore: certificateStore,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize server certificate manager: %v", err)
+	}
+
+	return serverCertManager, nil
+}
+
+// ApprApproveYurttunnelCSR starts a csr controller that automatically approves
+// the yurttunel related csr
+func ApproveYurttunnelCSR(
 	clientset kubernetes.Interface,
-	organization, commonName string) (certificate.Manager, error) {
-	return nil, errors.New("NOT IMPLEMENT YET")
+	stopCh <-chan struct{}) {
+	informerFactory := informers.
+		NewSharedInformerFactory(clientset, 10*time.Minute)
+	csrInformer := informerFactory.Certificates().V1beta1().
+		CertificateSigningRequests().Informer()
+	csrInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			approveYurttunnelCSR(obj, clientset)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			approveYurttunnelCSR(new, clientset)
+		},
+	})
+	go csrInformer.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, csrInformer.HasSynced) {
+		klog.Error("sync csr timeout")
+	}
+	<-stopCh
+}
+
+// approveYurttunnelCSR checks the csr status, if it is neither approved nor
+// denied, it will try to approve the csr.
+func approveYurttunnelCSR(obj interface{}, clientset kubernetes.Interface) {
+	csr := obj.(*certificates.CertificateSigningRequest)
+	if !isYurttunelCSR(csr) {
+		klog.Infof("csr(%s) is not Yurttunnel csr", csr.GetName())
+		return
+	}
+
+	approved, denied := checkCertApprovalCondition(&csr.Status)
+	if approved {
+		klog.Infof("csr(%s) is approved", csr.GetName())
+		return
+	}
+
+	if denied {
+		klog.Infof("csr(%s) is denied", csr.GetName())
+		return
+	}
+
+	// approve the edge tunnel server csr
+	csr.Status.Conditions = append(csr.Status.Conditions,
+		certificates.CertificateSigningRequestCondition{
+			Type:    certificates.CertificateApproved,
+			Reason:  "AutoApproved",
+			Message: "Auto approving edge tunnel server csr by edge tunnel server itself.",
+		})
+
+	csrClient := clientset.CertificatesV1beta1().CertificateSigningRequests()
+	result, err := csrClient.UpdateApproval(csr)
+	if err != nil {
+		if result == nil {
```

Comment:
if csr updates failed, maybe we need to add this csr in a backoff queue for csr update retry

Reply:
implement the YurttunelCSRApprover that approves all yurttunel related csr and has a backoff workqueue